rn/Teacher/ios/Teacher/SpeedGrader/SubmissionViewer.swift

```diff
@@ -20,25 +20,88 @@ import SwiftUI
+            }
+        case .discussion_topic, .online_quiz:
+            WebSession(url: submission.previewUrl) { url in
+                WebView(url: url)
+                    .onLink(handleLink)
+                    .onChangeSize(handleSize)
+                    .highPriorityGesture(DragGesture(), including: hasOverflow ? .all : .subviews)
+            }
+        case .media_recording:
+            VideoPlayer(url: submission.mediaComment?.url)
+        case .online_text_entry:
+            WebView(html: submission.body)
+                .onLink(handleLink)
+                .onChangeSize(handleSize)
+                .highPriorityGesture(DragGesture(), including: hasOverflow ? .all : .subviews)
+        case .online_upload:
+            let file = submission.attachments?.first { fileID == $0.id } ??
+                submission.attachments?.sorted(by: File.idCompare).first
+            if let file = file, let url = file.url, let previewURL = file.previewURL {
+                DocViewer(filename: file.filename, previewURL: previewURL, fallbackURL: url) {
+                    isPagingEnabled = !$0
+                }
+            } else if let id = file?.id {
+                FileViewer(fileID: id)
+            }
+        case .online_url:
+            URLSubmissionViewer(submission: submission)
+        default:
+            VStack {
                 Spacer()
-                Text("Submission Placeholder")
+                if assignment.submissionTypes.contains(.on_paper) {
+                    Text("This assignment only allows on-paper submissions.")
+                } else if assignment.submissionTypes.contains(.none) {
+                    Text("This assignment does not allow submissions.")
+                } else if submission.groupID != nil {
+                    Text("This group does not have a submission for this assignment.")
+                } else {
+                    Text("This student does not have a submission for this assignment.")
+                }
                 Spacer()
             }
-            Text(verbatim: "Attempt: #\(submission.attempt)")
-            Text(submission.submittedAt?.dateTimeString ?? "")
-            Text(verbatim: "Selected File: \(file?.id ?? "none")")
-            Spacer()
+                .font(.regular16).foregroundColor(.textDarkest)
+                .multilineTextAlignment(.center)
         }
     }
+
+    func handleLink(url: URL) -> Bool {
+        env.router.route(to: url, from: controller, options: .modal(embedInNav: true, addDoneButton: true))
+        return true
+    }
+
+    func openInSafari(url: URL) -> Bool {
+        env.loginDelegate?.openExternalURL(url)
+        return true
+    }
+
+    func handleSize(webView: Core.CoreWebView, height: CGFloat) {
+        guard self.webView.value != webView else { return }
+        let ref = WeakReference<Core.CoreWebView>()
+        ref.value = webView
+        self.webView = ref
```

Comment:
I'm surprised nothing is done with `height` here. I guess we don't need it because this web view does not grow.. Is this just to get a reference to the web view so that we can use it to check for overflow?

Reply:
Yeah, this is reusing the same callback to keep the webView so we can try to workaround the pages DragGesture crapping out. Without this stuff you can neither swipe to the next submission, nor scroll the web view horizontally.

I haven't given up on Pages yet, since it gives us some smooth nice animation, but perhaps later we may decide embedding each SubmissionGrader as a page in PagesViewController may eventually be better because we won't need these hacks.